make container runtime socket path and containerd ns configurable and change the socket volume path in pod

xlgao-zju opened the pull request:
<!--
Thank you for contributing to Chaos Mesh!

If you're unsure where to start, please refer to the contributing doc:

https://github.com/chaos-mesh/chaos-mesh/blob/master/CONTRIBUTING.md

If you still have questions, please let us know via issues.

Please follow the Title Formats below when you open a new PR:

1. module[, module2, module3]: what's changed
2. *: what's changed
-->

### What problem does this PR solve?
close #3072 

<!-- Uncomment this line if some issues to close -->
<!-- Close #<issue number> -->

### What's changed and how it works?
- [x] make container runtime socket path configurable
- [x] change the socket volume path in pod to fix #3072 
- [x] make container containerd ns configurable, to make all cr client configs configurable


<!-- Uncomment this line if this PR is associated with a proposal -->
<!-- Proposal: [name](url) -->

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [ ] release-2.1
  - [ ] release-2.0

### Checklist

Tests

<!-- Must include at least one of them. -->

- [x] Unit test
- [ ] E2E test
- [ ] No code
- [ ] Manual test (add steps below)

<!-- > steps: -->

Side effects

- [ ] Breaking backward compatibility

### Release note <!-- bugfixes or new feature need a release note -->

```text
Make container runtime socket path and containerd ns configurable, and change the socket volume path in pod
```

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

Signed-off-by: xianglingao <xianglingao@tencent.com>

xlgao-zju commented:
ping @FingerLeader for review too

fewdan commented:
Thanks for your contribution! It looks like CI is failing, specifically because of a "go vet" related issue.
You can execute "make check" before submitting PR.

fewdan commented:
BTW, UT failed. You can execute "make check" locally to check UT.

fewdan commented:
Unfortunately, it seems that CI failed again. @xlgao-zju

xlgao-zju commented:
@fewdan should I modify `install.sh` manually? or this shell is generated by `make check`?

since I get error message like this
```shell
Please make check before creating a PR
diff --git a/install.sh b/install.sh
index 2d98107..5bf60d4 100755
--- a/install.sh
+++ b/install.sh
@@ -1226,7 +1226,7 @@ spec:
             - !!str 31767
             - --pprof
             - --runtime-socket-path
-            -  /host-run/${socketName}
+            -  /host-run/docker.sock
           env:
             - name: TZ
               value: ${timezone}
@@ -1251,7 +1251,7 @@ spec:
       volumes:
         - name: socket-path
           hostPath:
-            path: ${socketDir}
+            path: /var/run
         - name: sys-path
           hostPath:
             path: /sys
```

fewdan commented:
@xlgao-zju 
try "./hack/update_install_script.sh"

xlgao-zju commented:
@fewdan yep, look like what I modified in `install.sh` should be generated by `./hack/update_install_script.sh`.
to make `install` right, I have modified the `./hack/update_install_script.sh`.

xlgao-zju commented:
@fewdan I taked to @FingerLeader offline about the failure of `Test install.sh`.  it may be a bug of the test itself. so this PR can not pass the test, until we fix the test. :(

fewdan commented:
> it may be a bug of the test itself. so this PR can not pass the test, until we fix the test. :(

@xlgao-zju Ah, so it is.

xlgao-zju commented:
@fewdan look like ci is happy now :)

YangKeao commented on helm/chaos-mesh/templates/chaos-daemon-daemonset.yaml:
```diff
@@ -176,7 +178,13 @@ spec:
       volumes:
         - name: socket-path
           hostPath:
-            path: {{ .Values.chaosDaemon.socketPath | default "/var/run/docker.sock" }}
```
If the user's container runtime has a different socket path (e.g. he set it manually), he will not be able to modify the socket path 🤔 , which is provided in the existed configuration. 

Or is there a way to be compatible with the old configuration? For example, automatically get the `dirname` of the path. @cwen0  @STRRL how do you think about it?

xlgao-zju replied:
> Or is there a way to be compatible with the old configuration? 
yes, we can use `dir` func in helm template to be compatible with the old configuration.

and at the same time, we can also provide a arg to let users customize socket dir, called `.Values.chaosDaemon.socketdir` or something.

xlgao-zju replied:
@YangKeao I use `dir` func in helm template to be compatible with the old configuration.

and also provide a arg called `.Values.chaosDaemon.socketDir` to let users customize socket dir.

PTAL

xlgao-zju commented:
all ci passed. ping @YangKeao @STRRL @FingerLeader for review :)

YangKeao commented on helm/chaos-mesh/templates/chaos-daemon-daemonset.yaml:
```diff
@@ -78,6 +78,14 @@ spec:
+          {{- end }}
+            - --runtime-socket-path
+          {{- if eq .Values.chaosDaemon.runtime "docker" }}
+            - /host-run/docker.sock
```
If the `socketPath` is set, we should also get the file name from it :thinking: .

xlgao-zju replied:
@YangKeao nice catch, updated

YangKeao reviewed (commented):
Rest LGTM

FingerLeader reviewed (approved):
LGTM

YangKeao reviewed (approved):
LGTM

YangKeao commented:
/merge
